### DISCUSSION TITLE
Make nested Zod object partial at only a specific level and add properties at specific level

### DISCUSSION THREAD

**Discussion by RenaudAubert** [colinhacks/zod#2935] [Q&A] [answered]
I'm currently working on an Express app and using Zod to validate an incoming request.
One "problem" I'm often facing is that the schema used to validate a body for a creation route for instance  `POST /contracts/:contractId` is often used again in a PATCH request except now all keys are optional.

I know Zod provide a `deepPartial` method but this not really what I want sometimes my schema contains 3 or 4 nested objects and I only want one level to be `partial`, moreover in Zod 3.22.4 `deepPartial` is deprecated.

What I'm currently doing is:

```javascript
const contractBodySchema = z.object({
    name: z.string(),
    nestedObject1: z.object({
        value1: z.number // this value should never be optional
    })
});

const contractCreationSchema = z.object({
    body: contractBodySchema
});

const contractUpdateSchema = z.object({
    body: contractBodySchema.partial()
});
```

It works fine but I was wondering if there was a simpler way to do this and maybe avoid creating the `contractBodySchema` and recreating multiple `z.object({ body: ... })`.

## Add property

Another problem I encountered is adding nested properties to an existing schema.

Let's say I have this schema

```javascript
const mySchema = z.object({
    value1: z.object({
        value2: z.object({
            value3: z.string()
        })
    })
})
```

How do I add `value4: z.number()` inside the `value2` Zod object?

Cheers.

**Comment by digisomni** [comment]
Has a solution been found? This kind of processing is critical for any real commercial applications, I'd expect there to be a well-supported way to handle this with Zod by now.

**Reply by JacobWeisenburger** [reply]
what specifically are you needing help with?

**Reply by RenaudAubert** [reply]
I did not found a solution for either of the problem I mentionned. For the first one, I'm just doing what is in the example but for the second one it is a bit more annoying

**Comment by JacobWeisenburger** [comment]
> How do I add `value4: z.number()` inside the value2 Zod object?

Is this what you are looking for?

```ts
const mySchema = z.object( {
    value1: z.object( {
        value2: z.object( {
            value3: z.string(),
            value4: z.number(),
        } )
    } )
} )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by RenaudAubert** [reply]
Hi!

Yes the validation you provided is what I would like to obtain from the original `mySchema`.
Right now I have to repeat each nesting level inside a new object and add the `value4: z.number(),` at the correct level but I was wondering if there was a better way maybe with `extend` or `merge`.

**Reply by JacobWeisenburger** [reply, marked as answer]
I don't think there is a good way to do what you are asking.

Here's the only way I can think of:

```ts
const mySchema = z.object( {
    value1: z.object( {
        value2: z.object( {
            value3: z.string(),
        } )
    } )
} )

const mySchema2 = mySchema.extend( {
    value1: mySchema.shape.value1.extend( {
        value2: mySchema.shape.value1.shape.value2.extend( {
            value4: z.number(),
        } )
    } )
} )

console.log( Object.keys( mySchema2.shape.value1.shape.value2.shape ) ) // [ 'value3', 'value4' ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by RenaudAubert** [reply]
I came up with the same solution at first so I guess there is no true way of achieving what I'm looking for. Anyway the syntax is so heavy that I chose to copy and paste the whole validation and add the nested field. It duplicates code but is easier to read.

I guess this also answers whether there is an easy way to make only some nested part of a validation partial.

Thanks a lot
